### PR TITLE
KEP-1012 pbft_operation_manager existential queries

### PR DESCRIPTION
**Pull request by paularchard**
Added various range queries to storage to provide support for finding/removing pbft operation state.

### CODE REVIEW

File: storage/rocksdb_storage.cpp [hunk 1]
```diff
@@ -385,3 +385,74 @@ rocksdb_storage::load_snapshot(const std::string& data)
 
     return false;
 }
+
+
+void
+rocksdb_storage::remove_range(const bzn::uuid_t& uuid, const std::string& first, const std::string& last)
+{
+    std::shared_lock<std::shared_mutex> lock(this->lock); // lock for read access
```
Comment (ebruck) [inline storage/rocksdb_storage.cpp]: you want write access which means a std::lock_guard should be used

File: storage/mem_storage.cpp [hunk 2]
```diff
@@ -264,3 +265,74 @@ mem_storage::load_snapshot(const std::string& data)
 
     return false;
 }
+
+void
+mem_storage::remove_range(const bzn::uuid_t& uuid, const std::string& first, const std::string& last)
+{
+    std::shared_lock<std::shared_mutex> lock(this->lock); // lock for read access
```
Comment (ebruck) [inline storage/mem_storage.cpp]: write lock required

File: storage/rocksdb_storage.cpp (hunk 1 above)
Comment (paularchard) [reply storage/rocksdb_storage.cpp]: Thanks. I didn't grok the lock semantics. Will fix.

File: storage/mem_storage.cpp (hunk 2 above)
Comment (paularchard) [reply storage/mem_storage.cpp]: ditto

File: pbft/operations/pbft_persistent_operation.cpp [hunk 3]
```diff
@@ -275,16 +328,76 @@ pbft_persistent_operation::get_preprepare() const
...
+        if (res)
+        {
+            uint64_t view;
+            uint64_t sequence;
+            bzn::hash_t hash;
+            if (parse_prefix(prefix, view, sequence, hash))
+            {
+                auto op = std::make_shared<pbft_persistent_operation>(storage, view, sequence, hash);
+                results.push_back(op);
+            }
```
[the remark this answers is not in the document]
Comment (paularchard) [reply pbft/operations/pbft_persistent_operation.cpp]: I'll add an error log message. You're correct, it should never fail. Alternatively I could assert the result.

File: pbft/operations/pbft_persistent_operation.hpp [hunk 4]
```diff
@@ -24,6 +24,7 @@ namespace bzn
     {
     public:
         pbft_persistent_operation(uint64_t view, uint64_t sequence, const bzn::hash_t& request_hash, std::shared_ptr<bzn::storage_base> storage, size_t );
+        pbft_persistent_operation(std::shared_ptr<bzn::storage_base> storage, uint64_t view, uint64_t sequence, const bzn::hash_t& request_hash);
```
[the remark this answers is not in the document]
Comment (paularchard) [reply pbft/operations/pbft_persistent_operation.hpp]: The primary constructor adds the operation to storage. It seemed strange to construct an operation from storage only to try and re-write it. I understand that write would come back with exists of course.  If you'd prefer to only have the single constructor and do the redundant store attempt I can change it.

File: pbft/operations/pbft_operation_manager.cpp [hunk 5]
```diff
@@ -108,26 +109,30 @@ pbft_operation_manager::prepared_operations_since(uint64_t sequence)
...
+        if (search == result.end() || result[op->get_sequence()]->get_view() < op->get_view())
         {
-            const auto search = result.find(op->get_sequence());
-            if (search == result.end() || result[op->get_sequence()]->get_view() < op->get_view())
-            {
-                result[op->get_sequence()] = op;
-            }
+            result[op->get_sequence()] = op;
         }
     };
```
[the remark this answers is not in the document]
Comment (paularchard) [reply pbft/operations/pbft_operation_manager.cpp]: I don't understand the comment. The code as written checks storage if we have it, or memory if we don't.

[the remark this answers is not in the document]
Comment (paularchard) [reply pbft/operations/pbft_operation_manager.cpp]: maybe the comment is the problem - // First, search through the operations we have in memory

**Pull request merged**